File: 20_Unity/Packages/io.github.hatayama.uloopmcp/TypeScriptServer~/src/utils/vibe-logger.ts
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { v4 as uuidv4 } from 'uuid';
import { OUTPUT_DIRECTORIES } from '../constants.js';

/**
 * AI-friendly structured logger for TypeScript MCP Server
 *
 * Design document reference: Packages/src/TypeScriptServer~/ARCHITECTURE.md
 *
 * Related classes:
 * - log-to-file.ts: Traditional file logger that this class replaces
 * - UnityDiscovery: Uses this logger for connection discovery tracking
 * - UnityClient: Uses this logger for connection event tracking
 *
 * Key features:
 * - Structured JSON logging with operation, context, correlation_id
 * - AI-friendly format for Claude Code analysis
 * - Automatic file rotation and memory management
 * - Correlation ID tracking for related operations
 */

interface VibeLogEntry {
  timestamp: string;
  level: string;
  operation: string;
  message: string;
  context?: unknown;
  correlation_id: string;
  source: string;
  human_note?: string;
  ai_todo?: string;
  environment: EnvironmentInfo;
}

interface EnvironmentInfo {
  node_version: string;
  platform: string;
  process_id: number;
  memory_usage: {
    rss: number;
    heapTotal: number;
    heapUsed: number;
  };
}

export class VibeLogger {
  private static readonly PROJECT_ROOT = VibeLogger.findUnityProjectRoot();
  private static readonly LOG_DIRECTORY = path.join(
    VibeLogger.PROJECT_ROOT,
    OUTPUT_DIRECTORIES.ROOT,
    OUTPUT_DIRECTORIES.VIBE_LOGS,
  );
  private static readonly LOG_FILE_PREFIX = 'ts_vibe';
  private static readonly MAX_FILE_SIZE_MB = 10;
  private static readonly MAX_MEMORY_LOGS = 1000;

  private static memoryLogs: VibeLogEntry[] = [];
  private static isDebugEnabled = process.env.MCP_DEBUG === 'true';

  /**
   * Log an info level message with structured context
   */
  static logInfo(
    operation: string,
    message: string,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace?: boolean,
  ): void {
    VibeLogger.log(
      'INFO',
      operation,
      message,
      context,
      correlationId,
      humanNote,
      aiTodo,
      includeStackTrace,
    );
  }

  /**
   * Log a warning level message with structured context and optional stack trace
   */
  static logWarning(
    operation: string,
    message: string,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace: boolean = true,
  ): void {
    VibeLogger.log(
      'WARNING',
      operation,
      message,
      context,
      correlationId,
      humanNote,
      aiTodo,
      includeStackTrace,
    );
  }

  /**
   * Log an error level message with structured context and optional stack trace
   */
  static logError(
    operation: string,
    message: string,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace: boolean = true,
  ): void {
    VibeLogger.log(
      'ERROR',
      operation,
      message,
      context,
      correlationId,
      humanNote,
      aiTodo,
      includeStackTrace,
    );
  }

  /**
   * Log a debug level message with structured context and optional stack trace
   */
  static logDebug(
    operation: string,
    message: string,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace?: boolean,
  ): void {
    VibeLogger.log(
      'DEBUG',
      operation,
      message,
      context,
      correlationId,
      humanNote,
      aiTodo,
      includeStackTrace,
    );
  }

  /**
   * Log an exception with structured context and optional additional stack trace
   */
  static logException(
    operation: string,
    error: Error,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace?: boolean,
  ): void {
    const exceptionContext = {
      original_context: context,
      exception: {
        name: error.name,
        message: error.message,
        stack: error.stack,
        cause: error.cause,
      },
    } as const;

    VibeLogger.log(
      'ERROR',
      operation,
      `Exception occurred: ${error.message}`,
      exceptionContext,
      correlationId,
      humanNote,
      aiTodo,
      includeStackTrace,
    );
  }

  /**
   * Generate a new correlation ID for tracking related operations
   */
  static generateCorrelationId(): string {
    const timestamp = new Date().toISOString().slice(11, 19).replace(/:/g, '');
    return `ts_${uuidv4().slice(0, 8)}_${timestamp}`;
  }

  /**
   * Get resolved Unity project root path.
   */
  static getProjectRoot(): string {
    return VibeLogger.PROJECT_ROOT;
  }

  /**
   * Get logs for AI analysis (formatted for Claude Code)
   * Output directory: {project_root}/.uloop/outputs/VibeLogs/
   */
  static getLogsForAi(operation?: string, correlationId?: string, maxCount: number = 100): string {
    let filteredLogs = [...VibeLogger.memoryLogs];

    if (operation) {
      filteredLogs = filteredLogs.filter((log) => log.operation.includes(operation));
    }

    if (correlationId) {
      filteredLogs = filteredLogs.filter((log) => log.correlation_id === correlationId);
    }

    if (filteredLogs.length > maxCount) {
      filteredLogs = filteredLogs.slice(-maxCount);
    }

    return JSON.stringify(filteredLogs, null, 2);
  }

  /**
   * Clear all memory logs
   */
  static clearMemoryLogs(): void {
    VibeLogger.memoryLogs = [];
  }

  /**
   * Write emergency log entry (for when main logging fails)
   * Static method to avoid circular dependency
   */
  private static writeEmergencyLog(emergencyEntry: Record<string, unknown>): void {
    try {
      // Security: Use safe path construction consistent with other logging
      const basePath = process.cwd();
      const sanitizedRoot = path.resolve(basePath, OUTPUT_DIRECTORIES.ROOT);
      if (!VibeLogger.validateWithin(basePath, sanitizedRoot)) {
        return; // Silent failure to prevent protocol interference
      }

      const emergencyLogDir = path.resolve(sanitizedRoot, 'EmergencyLogs');
      if (!VibeLogger.validateWithin(sanitizedRoot, emergencyLogDir)) {
        return; // Path traversal guard
      }

      // SECURITY: Path validated with validateWithin() above - safe from path traversal
      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
      fs.mkdirSync(emergencyLogDir, { recursive: true });

      const emergencyLogPath = path.resolve(emergencyLogDir, 'vibe-logger-emergency.log');
      if (!VibeLogger.validateWithin(emergencyLogDir, emergencyLogPath)) {
        return; // Path traversal guard
      }

      const emergencyLog = JSON.stringify(emergencyEntry) + '\n';
      // SECURITY: Path validated with validateWithin() above - safe from path traversal
      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
      fs.appendFileSync(emergencyLogPath, emergencyLog);
    } catch {
      // Silent failure to prevent MCP protocol interference
      // If even emergency logging fails, we cannot do anything more
    }
  }

  /**
   * Core logging method
   * Only logs when MCP_DEBUG environment variable is set to 'true'
   */
  private static log(
    level: string,
    operation: string,
    message: string,
    context?: unknown,
    correlationId?: string,
    humanNote?: string,
    aiTodo?: string,
    includeStackTrace?: boolean,
  ): void {
    // Only log when MCP_DEBUG is enabled
    if (!VibeLogger.isDebugEnabled) {
      return;
    }

    // Handle stack trace option
    let finalContext = context;
    if (includeStackTrace) {
      finalContext = {
        original_context: context,
        stack: VibeLogger.cleanStackTrace(new Error().stack),
      } as const;
    }

    const logEntry: VibeLogEntry = {
      timestamp: VibeLogger.formatTimestamp(),
      level,
      operation,
      message,
      context: VibeLogger.sanitizeContext(finalContext),
      correlation_id: correlationId || VibeLogger.generateCorrelationId(),
      source: 'TypeScript',
      human_note: humanNote,
      ai_todo: aiTodo,
      environment: VibeLogger.getEnvironmentInfo(),
    };

    // Add to memory logs
    VibeLogger.memoryLogs.push(logEntry);

    // Rotate memory logs if too many
    if (VibeLogger.memoryLogs.length > VibeLogger.MAX_MEMORY_LOGS) {
      VibeLogger.memoryLogs.shift();
    }

    // Save to file (fire and forget to avoid blocking)
    VibeLogger.saveLogToFile(logEntry).catch((error) => {
      // File logging failed - write to emergency log instead of console
      // Critical: No console output to avoid MCP protocol interference
      VibeLogger.writeEmergencyLog({
        timestamp: VibeLogger.formatTimestamp(),
        level: 'EMERGENCY',
        message: 'VibeLogger saveLogToFile failed',
        original_error: error instanceof Error ? error.message : String(error),
        original_log_entry: logEntry,
      });
    });

    // VibeLogger is designed for file output only - console output removed to prevent MCP protocol interference
  }

  /**
   * Validate file name to prevent dangerous characters
   */
  private static validateFileName(fileName: string): boolean {
    // Allow only alphanumeric characters, hyphens, underscores, and dots
    const safeFileNameRegex = /^[a-zA-Z0-9._-]+$/;
    return safeFileNameRegex.test(fileName) && !fileName.includes('..');
  }

  /**
   * Validate file path to prevent directory traversal attacks
   */
  private static validateFilePath(filePath: string): boolean {
    const normalizedPath = path.normalize(filePath);
    const logDirectory = path.normalize(VibeLogger.LOG_DIRECTORY);

    // Ensure the file path is within the log directory
    return normalizedPath.startsWith(logDirectory + path.sep) || normalizedPath === logDirectory;
  }

  /**
   * Validate that target path is within base directory
   */
  private static validateWithin(base: string, target: string): boolean {
    const resolvedBase = path.resolve(base);
    const resolvedTarget = path.resolve(target);
    return resolvedTarget.startsWith(resolvedBase);
  }

  /**
   * Safe wrapper for fs.existsSync with path validation
   */
  private static safeExistsSync(filePath: string): boolean {
    try {
      const absolutePath = path.resolve(filePath);
      const expectedDir = path.resolve(VibeLogger.LOG_DIRECTORY);

      // Validate path is within expected directory
      if (!VibeLogger.validateWithin(expectedDir, absolutePath)) {
        return false;
      }

      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
      return fs.existsSync(absolutePath);
    } catch {
      return false;
    }
  }

  /**
   * Prepare and validate log directory and file path
   */
  private static prepareLogFilePath(): string {
    // Validate log directory path for security
    const logDirectory = path.normalize(VibeLogger.LOG_DIRECTORY);
    if (!VibeLogger.validateFilePath(logDirectory)) {
      throw new Error('Invalid log directory path');
    }

    if (!VibeLogger.safeExistsSync(logDirectory)) {
      // Security: Ensure directory path is absolute and within expected bounds
      const absoluteLogDir = path.resolve(logDirectory);
      const expectedBaseDir = path.resolve(VibeLogger.PROJECT_ROOT);

      if (!VibeLogger.validateWithin(expectedBaseDir, absoluteLogDir)) {
        throw new Error('Log directory path escapes project root');
      }

      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
      fs.mkdirSync(absoluteLogDir, { recursive: true });
    }

    const fileName = `${VibeLogger.LOG_FILE_PREFIX}_${VibeLogger.formatDate()}.json`;

    // Validate file name for security
    if (!VibeLogger.validateFileName(fileName)) {
      throw new Error('Invalid file name detected');
    }

    const filePath = path.resolve(logDirectory, fileName);
    if (!VibeLogger.validateWithin(logDirectory, filePath)) {
      throw new Error('Resolved file path escapes the log directory');
    }

    // Validate the final file path for security
    if (!VibeLogger.validateFilePath(filePath)) {
      throw new Error('Invalid file path detected');
    }

    return filePath;
  }

  /**
   * Rotate log file if it exceeds maximum size
   */
  private static rotateLogFileIfNeeded(filePath: string): void {
    if (!VibeLogger.safeExistsSync(filePath)) {
      return;
    }

    // Security: Use absolute path for file operations
    const absoluteFilePath = path.resolve(filePath);
    // SECURITY: Using absolute path with security validation - safe from path traversal
    // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
    const stats = fs.statSync(absoluteFilePath);
    if (stats.size <= VibeLogger.MAX_FILE_SIZE_MB * 1024 * 1024) {
      return;
    }

    const logDirectory = path.dirname(filePath);
    const rotatedFileName = `${VibeLogger.LOG_FILE_PREFIX}_${VibeLogger.formatDateTime()}.json`;

    // Validate rotated file name for security
    if (!VibeLogger.validateFileName(rotatedFileName)) {
      throw new Error('Invalid rotated file name detected');
    }

    const rotatedFilePath = path.resolve(logDirectory, rotatedFileName);

    // Ensure rotated file path is within the allowed directory
    if (!VibeLogger.validateWithin(logDirectory, rotatedFilePath)) {
      throw new Error('Rotated file path escapes the allowed directory');
    }

    const sanitizedFilePath = path.resolve(logDirectory, path.basename(filePath));
    if (!VibeLogger.validateWithin(logDirectory, sanitizedFilePath)) {
      throw new Error('Original file path escapes the allowed directory');
    }

    // SECURITY: Both paths validated with validateWithin() above - safe from path traversal
    // eslint-disable-next-line security/detect-non-literal-fs-filename -- Both paths sanitized by validateWithin()
    fs.renameSync(sanitizedFilePath, rotatedFilePath);
  }

  /**
   * Write log to file with retry mechanism for concurrent access
   */
  private static async writeLogWithRetry(filePath: string, jsonLog: string): Promise<void> {
    const maxRetries = 3;
    const baseDelayMs = 50;

    // Security: Use absolute path and validate before writing
    const absoluteFilePath = path.resolve(filePath);
    const expectedLogDir = path.resolve(VibeLogger.LOG_DIRECTORY);

    if (!VibeLogger.validateWithin(expectedLogDir, absoluteFilePath)) {
      throw new Error('File path escapes log directory');
    }

    for (let retry = 0; retry < maxRetries; retry++) {
      try {
        // Use fs.promises.open for safer file handling

        // eslint-disable-next-line security/detect-non-literal-fs-filename
        const fileHandle = await fs.promises.open(absoluteFilePath, 'a');
        try {
          await fileHandle.writeFile(jsonLog, { encoding: 'utf8' });
        } finally {
          await fileHandle.close();
        }
        return; // Success - exit retry loop
      } catch (error: unknown) {
        if (VibeLogger.isFileSharingViolation(error) && retry < maxRetries - 1) {
          // Wait with exponential backoff for sharing violations
          const delayMs = baseDelayMs * Math.pow(2, retry);
          await VibeLogger.sleep(delayMs);
        } else {
          // For other errors or final retry, throw
          throw error;
        }
      }
    }
  }

  /**
   * Save log entry to file with retry mechanism for concurrent access
   */
  private static async saveLogToFile(logEntry: VibeLogEntry): Promise<void> {
    try {
      const filePath = VibeLogger.prepareLogFilePath();

      // Check file size and rotate if necessary
      VibeLogger.rotateLogFileIfNeeded(filePath);

      const jsonLog = JSON.stringify(logEntry) + '\n';
      await VibeLogger.writeLogWithRetry(filePath, jsonLog);
    } catch (error) {
      // File logging failed - try fallback logging
      await VibeLogger.tryFallbackLogging(logEntry, error);
    }
  }

  /**
   * Try fallback logging when main logging fails
   */
  private static async tryFallbackLogging(logEntry: VibeLogEntry, error: unknown): Promise<void> {
    try {
      // Security: Use safe path construction to prevent path traversal
      const basePath = process.cwd();
      const sanitizedRoot = path.resolve(basePath, OUTPUT_DIRECTORIES.ROOT);
      if (!VibeLogger.validateWithin(basePath, sanitizedRoot)) {
        throw new Error('Invalid OUTPUT_DIRECTORIES.ROOT path');
      }

      const safeLogDir = path.resolve(sanitizedRoot, 'FallbackLogs');
      if (!VibeLogger.validateWithin(sanitizedRoot, safeLogDir)) {
        throw new Error('Fallback log directory path traversal detected');
      }

      // SECURITY: Path validated with validateWithin() above - safe from path traversal
      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Path sanitized by validateWithin()
      fs.mkdirSync(safeLogDir, { recursive: true });

      // Security: Sanitize filename components to prevent path traversal
      const safeDate = VibeLogger.formatDateTime()
        .split(' ')[0]
        .replace(/[^0-9-]/g, '');
      const safeFilename = `${VibeLogger.LOG_FILE_PREFIX}_fallback_${safeDate}.json`;
      const safeFallbackPath = path.resolve(safeLogDir, safeFilename);

      // Security: Validate that the resolved file path is within our expected directory
      if (!VibeLogger.validateWithin(safeLogDir, safeFallbackPath)) {
        throw new Error('Invalid fallback log file path');
      }

      const fallbackEntry = {
        ...logEntry,
        fallback_reason: error instanceof Error ? error.message : String(error),
        original_timestamp: logEntry.timestamp,
      };

      const jsonLog = JSON.stringify(fallbackEntry) + '\n';

      // Use fs.promises.open for safer file writing

      // eslint-disable-next-line security/detect-non-literal-fs-filename
      const fileHandle = await fs.promises.open(safeFallbackPath, 'a');
      try {
        await fileHandle.writeFile(jsonLog, { encoding: 'utf8' });
      } finally {
        await fileHandle.close();
      }
    } catch (fallbackError) {
      // If even fallback fails, try to write to a last-resort emergency log file
      VibeLogger.writeEmergencyLog({
        timestamp: VibeLogger.formatTimestamp(),
        level: 'EMERGENCY',
        message: 'VibeLogger fallback failed',
        original_error: error instanceof Error ? error.message : String(error),
        fallback_error:
          fallbackError instanceof Error ? fallbackError.message : String(fallbackError),
        original_log_entry: logEntry,
      });
    }
  }

  /**
   * Check if error is a file sharing violation
   */
  private static isFileSharingViolation(error: unknown): boolean {
    if (!error) {
      return false;
    }

    // Check for common file sharing violation error codes
    const sharingViolationCodes = [
      'EBUSY', // Resource busy or locked
      'EACCES', // Permission denied (can indicate file in use)
      'EPERM', // Operation not permitted
      'EMFILE', // Too many open files
      'ENFILE', // File table overflow
    ];

    return (
      error &&
      typeof error === 'object' &&
      'code' in error &&
      typeof error.code === 'string' &&
      sharingViolationCodes.includes(error.code)
    );
  }

  /**
   * Asynchronous sleep function for retry delays
   */
  private static async sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Get current environment information
   */
  private static getEnvironmentInfo(): EnvironmentInfo {
    const memUsage = process.memoryUsage();
    return {
      node_version: process.version,
      platform: process.platform,
      process_id: process.pid,
      memory_usage: {
        rss: memUsage.rss,
        heapTotal: memUsage.heapTotal,
        heapUsed: memUsage.heapUsed,
      },
    };
  }

  /**
   * Sanitize context to prevent circular references
   */
  private static sanitizeContext(context: unknown): unknown {
    if (!context) {
      return undefined;
    }

    try {
      return JSON.parse(JSON.stringify(context));
    } catch {
      return {
        error: 'Failed to serialize context',
        original_type: typeof context,
        circular_reference: true,
      };
    }
  }

  /**
   * Format timestamp for log entries (local timezone)
   */
  private static formatTimestamp(): string {
    const now = new Date();
    const offset = now.getTimezoneOffset();
    const offsetHours = String(Math.floor(Math.abs(offset) / 60)).padStart(2, '0');
    const offsetMinutes = String(Math.abs(offset) % 60).padStart(2, '0');
    const offsetSign = offset <= 0 ? '+' : '-';

    return now.toISOString().replace('Z', `${offsetSign}${offsetHours}:${offsetMinutes}`);
  }

  /**
   * Format date for file naming (local timezone)
   */
  private static formatDate(): string {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    return `${year}${month}${day}`;
  }

  /**
   * Format datetime for file rotation
   */
  private static formatDateTime(): string {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    const hours = String(now.getHours()).padStart(2, '0');
    const minutes = String(now.getMinutes()).padStart(2, '0');
    const seconds = String(now.getSeconds()).padStart(2, '0');
    return `${year}${month}${day}_${hours}${minutes}${seconds}`;
  }

  /**
   * Clean stack trace by removing VibeLogger internal calls and Error prefix
   */
  private static cleanStackTrace(stack: string | undefined): string {
    if (!stack) {
      return '';
    }

    const lines = stack.split('\n');
    const cleanedLines: string[] = [];

    for (const line of lines) {
      // Skip the "Error" line at the beginning
      if (line.trim() === 'Error') {
        continue;
      }

      // Skip VibeLogger internal calls
      if (line.includes('vibe-logger.ts') || line.includes('VibeLogger')) {
        continue;
      }

      cleanedLines.push(line);
    }

    return cleanedLines.join('\n').trim();
  }

  /**
   * Find Unity project root by searching for Assets folder
   * Traverses parent directories from bundle location until Assets folder is found
   */
  private static findUnityProjectRoot(): string {
    let currentDir: string = path.dirname(fileURLToPath(import.meta.url));

    for (let i = 0; i < 20; i++) {
      const assetsPath: string = path.join(currentDir, 'Assets');
      // eslint-disable-next-line security/detect-non-literal-fs-filename -- Safe: path is constructed from module's own directory, only traverses upward
      if (fs.existsSync(assetsPath) && fs.statSync(assetsPath).isDirectory()) {
        return currentDir;
      }

      const parentDir: string = path.dirname(currentDir);
      if (parentDir === currentDir) {
        break;
      }
      currentDir = parentDir;
    }

    return path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../../..');
  }
}

File: 19_Unity/Packages/io.github.hatayama.uloopmcp/Editor/Shared/Logging/VibeLogger.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using Newtonsoft.Json;
using UnityEditor;
using UnityEngine;

namespace io.github.hatayama.uLoopMCP
{
    /// <summary>
    /// AI-friendly structured logger for Unity MCP
    /// 
    /// Design document reference: Packages/src/TypeScriptServer~/ARCHITECTURE.md
    /// 
    /// Related classes:
    /// - McpLogger: Traditional Unity logger that this class replaces
    /// - McpServerController: Uses this logger for domain reload tracking
    /// - UnityMcpServer: Uses this logger for connection event tracking
    /// 
    /// Key features:
    /// - Structured JSON logging with operation, context, correlation_id
    /// - AI-friendly format for Claude Code analysis
    /// - Automatic file rotation and memory management
    /// - Correlation ID tracking for related operations
    /// </summary>
    public static class VibeLogger
    {
        private static readonly string LOG_DIRECTORY = Path.Combine(Application.dataPath, "..", McpConstants.OUTPUT_ROOT_DIR, McpConstants.VIBE_LOGS_DIR);
        private static readonly string LOG_FILE_PREFIX = "unity_vibe";
        private static readonly int MAX_FILE_SIZE_MB = 10;
        private static readonly int MAX_MEMORY_LOGS = 1000;
        private static readonly int MAX_LOG_FILES = 3;
        
        private static readonly List<VibeLogEntry> memoryLogs = new();
        private static readonly object lockObject = new();
        private static bool hasCleanedUpOnStartup = false;
        
        [Serializable]
        public class VibeLogEntry
        {
            public string timestamp;
            public string level;
            public string operation;
            public string message;
            public object context;
            public string correlation_id;
            public string source;
            public string human_note;
            public string ai_todo;
            public string stack_trace;
            public EnvironmentInfo environment;
        }
        
        [Serializable]
        public class EnvironmentInfo
        {
            public string domain_reload_state;
        }
        
        /// <summary>
        /// Log an info level message with structured context
        /// Only logs when ULOOPENV_KEY_ULOOPMCP_DEBUG symbol is defined
        /// </summary>
        [Conditional(McpConstants.ENV_KEY_ULOOPMCP_DEBUG)]
        public static void LogInfo(string operation, string message, object context = null, 
                                  string correlationId = null, string humanNote = null, string aiTodo = null, bool includeStackTrace = false)
        {
            Log("INFO", operation, message, context, correlationId, humanNote, aiTodo, includeStackTrace);
        }
        
        /// <summary>
        /// Log a warning level message with structured context
        /// Only logs when ULOOPENV_KEY_ULOOPMCP_DEBUG symbol is defined
        /// </summary>
        [Conditional(McpConstants.ENV_KEY_ULOOPMCP_DEBUG)]
        public static void LogWarning(string operation, string message, object context = null, 
                                     string correlationId = null, string humanNote = null, string aiTodo = null, bool includeStackTrace = true)
        {
            Log("WARNING", operation, message, context, correlationId, humanNote, aiTodo, includeStackTrace);
        }
        
        /// <summary>
        /// Log an error level message with structured context
        /// Only logs when ULOOPENV_KEY_ULOOPMCP_DEBUG symbol is defined
        /// </summary>
        [Conditional(McpConstants.ENV_KEY_ULOOPMCP_DEBUG)]
        public static void LogError(string operation, string message, object context = null, 
                                   string correlationId = null, string humanNote = null, string aiTodo = null, bool includeStackTrace = true)
        {
            Log("ERROR", operation, message, context, correlationId, humanNote, aiTodo, includeStackTrace);
        }
        
        /// <summary>
        /// Log a debug level message with structured context
        /// Only logs when ULOOPENV_KEY_ULOOPMCP_DEBUG symbol is defined
        /// </summary>
        [Conditional(McpConstants.ENV_KEY_ULOOPMCP_DEBUG)]
        public static void LogDebug(string operation, string message, object context = null, 
                                   string correlationId = null, string humanNote = null, string aiTodo = null, bool includeStackTrace = false)
        {
            Log("DEBUG", operation, message, context, correlationId, humanNote, aiTodo, includeStackTrace);
        }
        
        /// <summary>
        /// Log an exception with structured context
        /// Only logs when ULOOPENV_KEY_ULOOPMCP_DEBUG symbol is defined
        /// </summary>
        [Conditional(McpConstants.ENV_KEY_ULOOPMCP_DEBUG)]
        public static void LogException(string operation, Exception exception, object context = null, 
                                       string correlationId = null, string humanNote = null, string aiTodo = null, bool includeStackTrace = true)
        {
            var exceptionContext = new Dictionary<string, object>();
            if (context != null)
            {
                exceptionContext["original_context"] = context;
            }
            
            exceptionContext["exception"] = new
            {
                type = exception.GetType().Name,
                message = exception.Message,
                stack_trace = exception.StackTrace,
                inner_exception = exception.InnerException?.Message
            };
            
            Log("ERROR", operation, $"Exception occurred: {exception.Message}", exceptionContext, 
                correlationId, humanNote, aiTodo, includeStackTrace);
        }
        
        /// <summary>
        /// Generate a new correlation ID for tracking related operations
        /// </summary>
        public static string GenerateCorrelationId()
        {
            return $"unity_{Guid.NewGuid().ToString("N")[..8]}_{DateTime.Now:HHmmss}";
        }
        
        /// <summary>
        /// Get logs for AI analysis (formatted for Claude Code)
        /// Output directory: {project_root}/.uloop/outputs/VibeLogs/
        /// </summary>
        public static string GetLogsForAi(string operation = null, string correlationId = null, int maxCount = 100)
        {
            lock (lockObject)
            {
                var filteredLogs = new List<VibeLogEntry>(memoryLogs);
                
                if (!string.IsNullOrEmpty(operation))
                {
                    filteredLogs = filteredLogs.FindAll(log => log.operation.Contains(operation));
                }
                
                if (!string.IsNullOrEmpty(correlationId))
                {
                    filteredLogs = filteredLogs.FindAll(log => log.correlation_id == correlationId);
                }
                
                if (filteredLogs.Count > maxCount)
                {
                    filteredLogs = filteredLogs.GetRange(filteredLogs.Count - maxCount, maxCount);
                }
                
                return JsonConvert.SerializeObject(filteredLogs, Formatting.Indented);
            }
        }
        
        /// <summary>
        /// Clear all memory logs
        /// </summary>
        public static void ClearMemoryLogs()
        {
            lock (lockObject)
            {
                memoryLogs.Clear();
            }
        }
        
        /// <summary>
        /// Core logging method
        /// </summary>
        private static void Log(string level, string operation, string message, object context,
                               string correlationId, string humanNote, string aiTodo, bool includeStackTrace = true)
        {
            var logEntry = new VibeLogEntry
            {
                timestamp = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss.fffzzz"),
                level = level,
                operation = operation,
                message = message,
                context = context,
                correlation_id = correlationId ?? GenerateCorrelationId(),
                source = "Unity",
                human_note = humanNote,
                ai_todo = aiTodo,
                stack_trace = includeStackTrace ? new StackTrace(true).ToString() : null,
                environment = GetEnvironmentInfo()
            };
            
            // Add to memory logs
            lock (lockObject)
            {
                memoryLogs.Add(logEntry);
                
                // Rotate memory logs if too many
                if (memoryLogs.Count > MAX_MEMORY_LOGS)
                {
                    memoryLogs.RemoveAt(0);
                }
            }
            
            // Save to file
            try
            {
                SaveLogToFile(logEntry);
            }
            catch (Exception ex)
            {
                // Fallback to Unity console if file logging fails
                UnityEngine.Debug.LogError($"[VibeLogger] Failed to save log to file: {ex.Message}");
                UnityEngine.Debug.Log($"[VibeLogger] {level} | {operation} | {message}");
            }
        }
        
        /// <summary>
        /// Save log entry to file with file locking for concurrent access
        /// </summary>
        private static void SaveLogToFile(VibeLogEntry logEntry)
        {
            if (!Directory.Exists(LOG_DIRECTORY))
            {
                Directory.CreateDirectory(LOG_DIRECTORY);
            }
            
            // Clean up old log files on first access only
            lock (lockObject)
            {
                if (!hasCleanedUpOnStartup)
                {
                    CleanupOldLogFiles();
                    hasCleanedUpOnStartup = true;
                }
            }
            
            string fileName = $"{LOG_FILE_PREFIX}_{DateTime.UtcNow:yyyyMMdd}.json";
            string filePath = Path.Combine(LOG_DIRECTORY, fileName);
            
            // Check file size and rotate if necessary
            if (File.Exists(filePath))
            {
                var fileInfo = new FileInfo(filePath);
                if (fileInfo.Length > MAX_FILE_SIZE_MB * 1024 * 1024)
                {
                    string rotatedFileName = $"{LOG_FILE_PREFIX}_{DateTime.UtcNow:yyyyMMdd_HHmmss}.json";
                    string rotatedFilePath = Path.Combine(LOG_DIRECTORY, rotatedFileName);
                    File.Move(filePath, rotatedFilePath);
                    
                    // Clean up old files after rotation
                    CleanupOldLogFiles();
                }
            }
            
            string jsonLog = JsonConvert.SerializeObject(logEntry) + "\n";
            
            // Use file locking with retry mechanism to handle concurrent access
            int maxRetries = 3;
            int retryDelayMs = 50;
            
            for (int retry = 0; retry < maxRetries; retry++)
            {
                try
                {
                    using (var fileStream = new FileStream(filePath, FileMode.Append, FileAccess.Write, FileShare.Read))
                    using (var writer = new StreamWriter(fileStream))
                    {
                        writer.Write(jsonLog);
                        writer.Flush();
                        return; // Success - exit retry loop
                    }
                }
                catch (IOException ex) when (IsFileSharingViolation(ex) && retry < maxRetries - 1)
                {
                    // Wait and retry for sharing violations
                    System.Threading.Thread.Sleep(retryDelayMs * (retry + 1));
                }
                catch (Exception ex)
                {
                    // For other exceptions, throw immediately
                    throw new InvalidOperationException($"Failed to write VibeLogger entry to file: {ex.Message}", ex);
                }
            }
            
            // If all retries failed, throw with sharing violation context
            throw new InvalidOperationException($"Failed to write VibeLogger entry after {maxRetries} retries due to file sharing violations");
        }
        
        /// <summary>
        /// Check if exception is a file sharing violation
        /// </summary>
        private static bool IsFileSharingViolation(IOException ex)
        {
            // ERROR_SHARING_VIOLATION (0x80070020)
            const int ERROR_SHARING_VIOLATION = unchecked((int)0x80070020);
            return ex.HResult == ERROR_SHARING_VIOLATION;
        }
        
        /// <summary>
        /// Clean up old log files, keeping only the most recent MAX_LOG_FILES
        /// </summary>
        private static void CleanupOldLogFiles()
        {
            try
            {
                if (!Directory.Exists(LOG_DIRECTORY))
                    return;
                    
                // Get all vibe log files, sorted by creation time (newest first)
                var logFiles = Directory.GetFiles(LOG_DIRECTORY, $"{LOG_FILE_PREFIX}_*.json")
                    .Select(file => new FileInfo(file))
                    .OrderByDescending(file => file.CreationTime)
                    .ToArray();
                    
                // Delete files beyond the limit
                for (int i = MAX_LOG_FILES; i < logFiles.Length; i++)
                {
                    try
                    {
                        logFiles[i].Delete();
                    }
                    catch (Exception ex)
                    {
                        // Log deletion failure but don't crash
                        UnityEngine.Debug.LogWarning($"[VibeLogger] Failed to delete old log file {logFiles[i].Name}: {ex.Message}");
                    }
                }
            }
            catch (Exception ex)
            {
                UnityEngine.Debug.LogWarning($"[VibeLogger] Failed to cleanup old log files: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Get current environment information
        /// </summary>
        private static EnvironmentInfo GetEnvironmentInfo()
        {
            bool isDomainReloadInProgress = EditorApplication.isCompiling || DomainReloadStateRegistry.IsDomainReloadInProgress();

            return new EnvironmentInfo
            {
                domain_reload_state = isDomainReloadInProgress ? "InProgress" : "Idle"
            };
        }
    }
}
